csharp/ql/src/semmle/code/csharp/frameworks/NHibernate.qll

```diff
@@ -0,0 +1,115 @@
+import csharp
+private import semmle.code.csharp.frameworks.System
+private import semmle.code.csharp.frameworks.system.Collections
+private import semmle.code.csharp.frameworks.Sql
+
+module NHibernate {
+  /** A class that is mapped to the database. */
+  abstract class MappedClass extends Class { }
+
+  /** The interface `NHibernamte.ISession`. */
+  class ISessionInterface extends Interface {
+    ISessionInterface() { this.hasQualifiedName("NHibernate.ISession") }
+
+    /** Gets a parameter that uses a mapped object. */
+    Parameter getAMappedObjectParameter() {
+      exists(Callable c |
+        result.getType() instanceof ObjectType and
+        c = this.getAMethod() and
+        result = c.getAParameter() and
+        result.getName() = "obj"
+      )
+    }
+
+    /** Gets a type parameter that specifies a mapped class. */
+    TypeParameter getAMappedObjectTp() {
+      exists(string methodName |
+        methodName = "Load"
+        or
+        methodName = "Merge"
+        or
+        methodName = "Get"
+        or
+        methodName = "Query"
+      |
+        result = this.getAMethod(methodName).(UnboundGenericMethod).getTypeParameter(0)
+      )
+    }
+  }
+
+  /** A mapped class that is mapped because it is used as a type argument. */
+  private class MappedByTypeArgument extends MappedClass {
+    UnboundGeneric gen;
+
+    MappedByTypeArgument() {
+      this = any(ISessionInterface si).getAMappedObjectTp().getASuppliedType()
+    }
+  }
+
+  /** A mapped class that is mapped because it is passed as a parameter. */
+  private class MappedByParam extends MappedClass {
+    MappedByParam() {
+      exists(ISessionInterface si, Expr e, MethodCall c, Parameter p |
+        p = si.getAMappedObjectParameter() and
+        e = c.getArgumentForParameter(p) and
+        this = e.getType()
+      ) and
+      not this instanceof ObjectType and
+      not this.getABaseInterface*() instanceof SystemCollectionsIEnumerableInterface and
```

Comment:
why?

Reply:
Sometimes, the argument could be a `List` or a `Dictionary` which isn't supposed to be mapped. This happened on the NHibernate-Core codebase anyway.